experimental/dds/tree2/src/shared-tree/sharedTreeChangeFamily.ts

```diff
@@ -0,0 +1,167 @@
+import { makeSharedTreeChangeCodecFamily } from "./sharedTreeChangeCodecs";
+import { SharedTreeChange } from "./sharedTreeChangeTypes";
+import { SharedTreeEditBuilder } from "./sharedTreeEditBuilder";
+
+/**
+ * Implementation of {@link ChangeFamily} that combines edits to fields and schema changes.
+ *
+ * @sealed
+ */
+export class SharedTreeChangeFamily
+	implements
+		ChangeFamily<SharedTreeEditBuilder, SharedTreeChange>,
+		ChangeRebaser<SharedTreeChange>
+{
+	public static emptyChange: SharedTreeChange = {
+		changes: [],
+	};
+
+	public readonly codecs: ICodecFamily<SharedTreeChange>;
+	private readonly modularChangeFamily: ModularChangeFamily;
+
+	public constructor(codecOptions: ICodecOptions) {
+		this.modularChangeFamily = new ModularChangeFamily(fieldKinds, codecOptions);
+		this.codecs = makeSharedTreeChangeCodecFamily(
+			this.modularChangeFamily.latestCodec,
+			codecOptions,
+		);
+	}
+
+	public buildEditor(changeReceiver: (change: SharedTreeChange) => void): SharedTreeEditBuilder {
+		return new SharedTreeEditBuilder(this.modularChangeFamily, changeReceiver);
+	}
+
+	public compose(changes: TaggedChange<SharedTreeChange>[]): SharedTreeChange {
+		const newChanges: Mutable<SharedTreeChange["changes"]> = [];
+		const dataChangeRun: TaggedChange<ModularChangeset>[] = [];
+
+		const flushDataChangeRun = (): void => {
+			if (dataChangeRun.length > 0) {
+				newChanges.push({
+					type: "data",
+					innerChange: this.modularChangeFamily.compose(dataChangeRun),
+				});
+				dataChangeRun.length = 0;
+			}
+		};
+
+		for (const topChange of changes) {
+			for (const change of topChange.change.changes) {
+				if (change.type === "schema") {
+					flushDataChangeRun();
+					newChanges.push(change);
+				} else {
+					dataChangeRun.push(mapTaggedChange(topChange, change.innerChange));
+				}
+			}
+		}
+
+		flushDataChangeRun();
+		return { changes: newChanges };
+	}
+
+	public invert(change: TaggedChange<SharedTreeChange>, isRollback: boolean): SharedTreeChange {
+		const invertInnerChange: (
+			innerChange: SharedTreeChange["changes"][number],
+		) => SharedTreeChange["changes"][number] = (innerChange) => {
+			switch (innerChange.type) {
+				case "data":
+					return {
+						type: "data",
+						innerChange: this.modularChangeFamily.invert(
+							mapTaggedChange(change, innerChange.innerChange),
+							isRollback,
+						),
+					};
+				case "schema": {
+					return {
+						type: "schema",
+						innerChange: {
+							schema: {
+								new: innerChange.innerChange.schema.old,
+								old: innerChange.innerChange.schema.new,
+							},
+						},
+					};
+				}
+				default:
+					fail("Unknown SharedTree change type.");
+			}
+		};
+		return {
+			changes: change.change.changes.map(invertInnerChange).reverse(),
+		};
+	}
+
+	public rebase(
+		change: SharedTreeChange,
+		over: TaggedChange<SharedTreeChange>,
+		revisionMetadata: RevisionMetadataSource,
+	): SharedTreeChange {
+		/**
```

Comment:
This comment ends up being associated with `hasSchemaChange` which seems a little arbitrary.
Perhaps a `//`-based comment would make more sense here.